primitives/header-chain/tests/implementation_match.rs

```diff
@@ -0,0 +1,317 @@
+// Copyright 2020-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tests inside this module are made to ensure that our custom justification verification
+//! implementation works exactly as `fn finality_grandpa::validate_commit`.
+//!
+//! Some of tests in this module may partially duplicate tests from `justification.rs`,
+//! but their purpose is different.
+
+use assert_matches::assert_matches;
+use bp_header_chain::justification::{verify_justification, Error, GrandpaJustification};
+use bp_test_utils::{
+	header_id, make_justification_for_header, signed_precommit, test_header, Account, JustificationGeneratorParams,
+	ALICE, BOB, CHARLIE, DAVE, EVE, TEST_GRANDPA_SET_ID,
+};
+use finality_grandpa::voter_set::VoterSet;
+use sp_finality_grandpa::{AuthorityId, AuthorityWeight};
+use sp_runtime::traits::Header as HeaderT;
+
+type TestHeader = sp_runtime::testing::Header;
+type TestHash = <TestHeader as HeaderT>::Hash;
+type TestNumber = <TestHeader as HeaderT>::Number;
+
+/// Implementation of `finality_grandpa::Chain` that is used in tests.
+struct AncestryChain(bp_header_chain::justification::AncestryChain<TestHeader>);
+
+impl AncestryChain {
+	fn new(ancestry: &[TestHeader]) -> Self {
+		Self(bp_header_chain::justification::AncestryChain::new(ancestry))
+	}
+}
+
+impl finality_grandpa::Chain<TestHash, TestNumber> for AncestryChain {
```

Comment:
If this implementation isn't up to date with what's in `finality-grandpa` our checks could fall out of line with what's considered "valid", is that right?

Reply:
There's no implementation of `Chain` in `finality-grandpa`. Is that what you mean?

Reply:
Sorry, I meant in [Substrate](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/justification.rs#L224), not `finality-grandpa`

Reply:
In theory - yes. But in practice I don't think that the meaning of `fn ancestry()` would change :) There's also a lot of other places where we may diverge. But again - I'm not sure whether it is critical issue, or not.

For example - AFAIU the valid justification is justification that presents 2/3+1 signatures for descendants of finality target. But what if e.g. someone would add **additional** invalid signature to the justification. So if validators set is 10 validators, we need 7 correct signatures to prove finality of header. What if there's eighth signature in justification and this signature is invalid? Right now both original (Substrate + finality-grandpa) and our implementations would reject such justification. But even though this justification contains invalid data, it still proves finality (it has 7 valid signatures out of 10). If this will ever change in original implementation (it'll ignore invalid vote and accept such justification), the node would accept such justification and the relayer that is connected to this node will try to submit such justification to the runtime. This transaction will fail (becaue) and it'll keep retrying forever.

This ^^^ is bad, so I've decided to go with same implementations for now.

TLDR: having different implementations of `finality_grandpa::Chain` is not the only thing that may lead to the failure + it is unlikely to diverge imo.